crates/matrix-sdk-base/src/sliding_sync/mod.rs

```diff
@@ -555,20 +576,35 @@ impl BaseClient {
             );
             let mut room_info = room.clone_info();
 
-            // We don't actually know what events are inside invite_state. In theory, they
-            // might not contain an m.room.member event, or they might set the
-            // membership to something other than invite. This would be very
-            // weird behaviour by the server, because invite_state is supposed
-            // to contain an m.room.member. We will call handle_invited_state, which will
-            // reflect any information found in the real events inside
-            // invite_state, but we default to considering this room invited
-            // simply because invite_state exists. This is needed in the normal
-            // case, because the sliding sync server tries to send minimal state,
-            // meaning that we normally actually just receive {"type": "m.room.member"} with
-            // no content at all.
-            room_info.mark_as_invited();
+            // We need to find the membership event since it could be for either an invited
+            // or knocked room
+            let membership_event_content = invite_state.iter().find_map(|raw| {
+                raw.deserialize().ok().iter().find_map(|e| {
```

Comment:
So I'm not sure we're looking in the right place here: the `required_state` events have already been deserialized and passed as a function parameter called `state_events`. Might be worth checking against a real-world Synapse, but I'd expect the membership event to be there instead. If that's true, then we don't need the deserialization step here, and this solution feels better than the second one IMO (with the previous sync protocol, knocked rooms would appear in the knocked subset of the response, which is nice and simple — I'd rather not change the room state based on extra other information).

Reply:
Invite and Knock rooms use `invite_state` instead of `required_state`, I suspect their required_state would be empty, but I can double check.

Reply:
Yes, there's no `required_state` for knocked rooms:

```json
"!luXevcpKTZauFIpPnL:matrix.org": {
  "bump_stamp": 21538303,
  "notification_count": 0,
  "highlight_count": 0,
  "initial": true,
  "invite_state": [
    {
      "content": {
        "creator": "@jorgem_test:matrix.org",
        "room_version": "7"
      },
      "sender": "@jorgem_test:matrix.org",
      "state_key": "",
      "type": "m.room.create"
    },
    {
      "content": {
        "join_rule": "knock"
      },
      "sender": "@jorgem_test:matrix.org",
      "state_key": "",
      "type": "m.room.join_rules"
    },
    {
      "content": {
        "name": "Knocking room test"
      },
      "sender": "@jorgem_test:matrix.org",
      "state_key": "",
      "type": "m.room.name"
    },
    {
      "content": {
        "algorithm": "m.megolm.v1.aes-sha2"
      },
      "sender": "@jorgem_test:matrix.org",
      "state_key": "",
      "type": "m.room.encryption"
    },
    {
      "type": "m.room.member",
      "state_key": "@jorgem:element.io",
      "content": {
        "avatar_url": "mxc://element.io/75d9a551895233d284aff390d2086814a2f7e875",
        "displayname": "Jorge",
        "membership": "knock"
      },
      "sender": "@jorgem:element.io"
    }
  ]
}
```

Reply:
Another option would be to handle the stripped state events in `invite_state` as part of the membership in this same method, but I think that's not possible because we need the `push_rules` too which are processed after this.

Reply:
We could also deserialize the stripped state events in `process_sliding_sync_room` and pass it down both functions, this one and `handle_invited_state`, so they'd be deserialized just once.

Reply:
I tried this approach ⬆️  in https://github.com/matrix-org/matrix-rust-sdk/pull/4128/commits/0b1ef5b3fe107e219fb93278b973affdb94b949c